src/kernel/pmm.zig

```diff
@@ -0,0 +1,175 @@
+const log = @import("log.zig");
+const MEMORY_AVAILABLE = @import("multiboot.zig").MULTIBOOT_MEMORY_AVAILABLE;
+const Bitmap = @import("bitmap.zig").Bitmap;
+
+const PmmBitmap = Bitmap(u32);
+
+/// The possible errors thrown by bitmap functions
+const Error = error{
+    /// The address given hasn't been allocated
+    NotAllocated,
+};
+
+/// The size of memory associated with each bitmap entry
+const BLOCK_SIZE = arch.MEMORY_BLOCK_SIZE;
+
+var bitmap: PmmBitmap = undefined;
+
+///
+/// Set the bitmap entry for an address as occupied
+///
+/// Arguments:
+///     IN addr: usize - The address.
+///
+/// Error: PmmBitmap.Error.
+///     *: See PmmBitmap.setEntry.
+///
+fn setAddr(addr: usize) PmmBitmap.Error!void {
+    try bitmap.setEntry(@intCast(u32, addr / BLOCK_SIZE));
+}
+
+///
+/// Check if an address is set as occupied.
+///
+/// Arguments:
+///     IN addr: usize - The address to check.
+///
+/// Return: True if occupied, else false.
+///
+/// Error: PmmBitmap.Error.
+///     *: See PmmBitmap.setEntry.
+///
+fn isSet(addr: usize) PmmBitmap.Error!bool {
+    return bitmap.isSet(@intCast(u32, addr / BLOCK_SIZE));
+}
+
+///
+/// Intiialise the physical memory manager and set all unavailable regions as occupied (those from the memory map and those from the linker symbols).
+///
+/// Arguments:
+///     IN mem: *const MemProfile - The system's memory profile.
+///     IN allocator: *std.mem.Allocator - The allocator to use to allocate the bitmaps.
+///
+pub fn init(mem: *const MemProfile, allocator: *std.mem.Allocator) void {
+    log.logInfo("Init pmm\n");
+    bitmap = PmmBitmap.init(mem.mem_kb * 1024 / BLOCK_SIZE, allocator) catch @panic("Bitmap allocation failed");
+
+    // Occupy the regions of memory that the memory map describes as reserved
+    for (mem.mem_map) |entry| {
+        if (entry.@"type" != MEMORY_AVAILABLE) {
+            var addr = std.mem.alignBackward(@intCast(usize, entry.addr), BLOCK_SIZE);
+            var end = @intCast(usize, entry.addr + (entry.len - 1));
+            // If the end address can be aligned without overflowing then align it
+            if (end <= std.math.maxInt(usize) - BLOCK_SIZE)
+                end = std.mem.alignForward(end, BLOCK_SIZE);
+            while (addr < end) : (addr += BLOCK_SIZE) {
+                setAddr(addr) catch |e| switch (e) {
+                    // We can ignore out of bounds errors as the memory won't be available anyway
+                    PmmBitmap.Error.OutOfBounds => break,
+                    else => panic(@errorReturnTrace(), "Failed setting address 0x{x} from memory map as occupied: {}", addr, e),
+                };
+            }
+        }
+    }
+    // Occupy kernel memory
+    var addr = std.mem.alignBackward(@ptrToInt(mem.physaddr_start), BLOCK_SIZE);
+    while (addr < @ptrToInt(mem.physaddr_end)) : (addr += BLOCK_SIZE) {
+        setAddr(addr) catch |e| switch (e) {
+            PmmBitmap.Error.OutOfBounds => break,
```

Comment:
Surely it will be  problem if we can't set the kernel address space, so should panic?

Reply:
Yeah we should :+1: